improve policy verification

flavio opened the pull request:
> This PR supersedes https://github.com/kubewarden/policy-server/pull/214

- feat: Add keyless verification by supporting verification-config.yml
- feat: fetch sigstore data from remote TUF repo

Fixes https://github.com/kubewarden/policy-server/issues/211

ereslibre commented on src/main.rs:
```diff
@@ -179,60 +171,49 @@ fn main() -> Result<()> {
+                        .map_err(|e| {
+                            fatal_error(format!("Policy '{}' cannot be verified: {:?}", name, e))
+                        })
+                        .unwrap(),
```
In case of more than one offender this is only reporting the first one and exiting, right? I think it would be nice to report all offenders at once.

viccuad replied:
No, it prints all the constraints from the verification_config that are not satisfied. E.g:

https://github.com/kubewarden/policy-fetcher/blob/6fe85ba4f7a64e60171aead5511cea34d78d261d/src/verify/mod.rs#L526-L539

raulcabello replied:
yes, but it will exit (fatal_error) as soon as one policy fails without evaluating the others policies pending in the loop

raulcabello reviewed (commented):
Are we going to allow users to use their own Fulcio certificate as we do in kwctl? I see we always fetch the code from TUF `FulcioAndRekorData::FromTufRepository` , but there is no way to use your own certificate `FulcioAndRekorData::FromCustomData`

viccuad commented on src/cli.rs:
```diff
@@ -183,28 +184,16 @@ pub(crate) fn policies(matches: &clap::ArgMatches) -> Result<HashMap<String, Pol
-pub(crate) fn verification_settings(matches: &clap::ArgMatches) -> Result<VerificationSettings> {
-    if matches.value_of("verification-path").is_none() {
-        return Err(anyhow!(
+pub(crate) fn verification_settings(
```
Nit:

```suggestion
pub(crate) fn verification_config(
```
to match policy-fecther (and later variable)

viccuad reviewed (approved):
LGTM, besides small nit

flavio commented:
I've updated the PR to reflect the changes you requested. I've converted the PR to be a draft because the policy-fetcher and policy-evalutator dependencies need to be updated. This is currently blocked by https://github.com/kubewarden/policy-fetcher/pull/69

flavio commented:
> Are we going to allow users to use their own Fulcio certificate as we do in kwctl? I see we always fetch the code from TUF `FulcioAndRekorData::FromTufRepository` , but there is no way to use your own certificate `FulcioAndRekorData::FromCustomData`

Yes, this isn't supported yet. I would add that via another PR.  This is something I don't see as super important right now for Policy Server.

Moreover, there's something that needs to be sorted out: we must provide a way to specify multiple fulcio certs via the cli flag. However, I have no idea about how this could be done via environment flags. Right now our controller uses environment variables as a way to configure the behaviour of policy-server. I don't know how we could expose that...

flavio commented:
This can be reviewed now. I'm working on adding some unit tests for the downloader code

flavio commented:
The unit tests are ready

flavio commented:
This is now building fine, because all the dependencies are now available :partying_face:
